ecommerce/api.py

```diff
@@ -103,3 +111,184 @@ def make_reference_id(order):
     return (
         f"{_REFERENCE_NUMBER_PREFIX}{settings.CYBERSOURCE_REFERENCE_PREFIX}-{order.id}"
     )
+
+
+def get_valid_coupon_versions(coupons, user, auto_only=False):
+    """
+    Given a list of coupon ids, determine which of them are valid based on invoice version dates and redemptions.
+
+    Args:
+        coupons (QuerySet of coupon ids): List of coupons to filter for validity
+        auto_only (bool): Whether or not to filter by automatic=True
+
+    Returns:
+        list of CouponVersion: CouponVersion objects sorted by discount, highest first.
+    """
+    valid_coupons = []
+    now = now_in_utc()
+
+    # Get the ids of the latest coupon versions
+    cv_latest = (
+        CouponVersion.objects.select_related()
+        .filter(coupon__in=coupons)
+        .order_by("coupon", "-created_on")
+        .distinct("coupon")
+        .values_list("pk")
+    )
+
+    # filter by expiration and activation dates
+    query = (
```

Comment:
Yes, initially this was part of the same query but I was getting the wrong CouponVersion back, because the distinct was being applied after filtering on dates.